TACC/protx-dashboard
# Task/COOKS-260-reactivate-frontend-linting

## Overview: ##
**Reactivate Frontend Linting and Formatting**
- ESLint for Javascript and Typescript
- Stylelint for CSS
- Prettier for formatting

## Related Jira tickets: ##

* [COOKS-260](https://jira.tacc.utexas.edu/browse/COOKS-123)

## Summary of Changes: ##
Installed the latest versions of ESLint, Stylelint, and Prettier.

`main.yml` (for Github actions) updated for client side linting -- necessary code was copied directly from Core-Portal's`main.yml` file.

Content in the configuration files (e.g. `eslintrc.json`, `stylelintrc.js`, `.prettierrc`) were largely based on Core-Portal's files of the same type.

The latest version of Stylelint has new rules; I've turned some of them off for now in `.stylelintrc.js` to match Core-Portal's Stylelint behavior. New deactivated rules: `selector-not-notation`, `color-function-notation`, `alpha-value-notation`, `import-notation`.

I also installed an scss plug-in for Stylelint ([stylelint-scss](https://www.npmjs.com/package/stylelint-scss)). Primary reason was to get rid of the annoying Stylelint warnings about not having an scss plug-in, secondary reason was to enable scss linting :smile:.

Below are changes based on **linting** errors that popped up:
- ESLint: change `class` to `className` in a div in the `AnalyticsStateDistribution` component. 
- ESLint: removed `styleName` prop from two `input` components in `DisplaySelectors.jsx` component. `styleName` is not a valid prop for the `input` component.
- Stylelint: Note CSS changes in the following files: `ChartInstructions.css`, `MainMap.css`, `MainPlot.css`, `App.css`

## Testing Steps: ##
There should be no noticeable differences between dev and production.
1. Go to https://cep.test/protx/dash/demographics and compare to production ([https://ccprotx.org/demographics/](https://ccprotx.org/demographics/)).
2. Go to https://cep.test/protx/dash/analytics and compare to production ([https://ccprotx.org/analytics/](https://ccprotx.org/analytics/).

## UI Photos:

## Notes: ##
Run `npm ci` in `protx-dashboard/protx-client` to get the new linting & formatting packages.

New scripts have been added to package.json with the same commands as Core-Portal:

- `npm run lint` to run ESLint, Stylelint, and Prettier check
- `npm run lint:css` to run Stylelint only (CSS)
- `npm run lint:js` to run ESLint only (Javascript)
- `npm run prettier:check` to **check** Prettier formatting
- `npm run prettier:fix` to **apply** Prettier formatting
